### PR TITLE
Finalize RC4 bootstrap keyring without activating strict signing

### PR DESCRIPTION
**Pull request by malik-na**
Finalize the RC4 bootstrap keyring before enabling signed RC5. Ship only primary `FBD6874D423C418DDB6D143EECE19CDDE306DBD2` in `omarchy-mac-keyring` version `20260914-2`, matching https://github.com/omarchy-mac/omarchy-pkgs-aarch64/pull/28. Fresh installation and private channel trust use the new primary; previously installed old and unrelated trust is preserved.

The version stays `4.0.3rc4`. The four repository templates and channel policy renderer retain the reviewed RC4 behavior, including preservation of stricter existing policies. The strict RC5 migration is absent.

Review cleanup at `07753c84583aee5f112adc3d74bacdb300a7be0d`:

- The pending bootstrap migration checks the installed keyring is at least `20260914-2` before its explicit trust population and reports package/version, population and fingerprint failures. The existing package helper still receives a plain package name; its skip-success case is checked explicitly.
- Remove redundant `1789390468.sh` and its invalid top-level `return`. Retain `1789407944.sh` for clients with completed earlier markers. Failed migrations remain pending and stop the queue.
- Replace the native test's unpublished Git-commit dependency with checked-in public package fixtures, checksums and provenance. Use a neutral, filesystem-checked scratch directory and explicitly enable the test in the existing disposable ARM guest.
- Document key replacement, retained old trust, package-before-migration ordering and the strict-policy canary limitation. Correct the misplaced preflight comment.

The fingerprint-constant refactor and dependency-parser simplification/naming suggestions remain deferred. All nine review conversations have responses describing what was changed, what remains unverified, or why an item was deferred; replies do not imply reviewer approval or conversation resolution.

Validation at the cleanup:

- Six focused migration/trust and CI contract/source-isolation suites pass. Tests use the real migration runner and package helpers with mocked package/privilege boundaries, including unavailable and outdated packages, successful repair, failure diagnostics and marker behavior.
- Three actual native package scenarios pass locally from a source-only directory without `.git`: uninitialized trust, fresh installation, and prior-version upgrade preserving old/unrelated trust.
- ShellCheck and syntax checks pass. Independent safety review found no blocking issue in the migration or native-test/CI diff.
- Full local QA passed 119 CLI assertions and 293/294 shell files with 3,420 passing shell assertions. The sole `update-lock` failure was previously reproduced on the unchanged baseline; this is not reported as a complete local-suite pass.
- Both hosted standard CI runs passed: https://github.com/omacom/omarchy-mac/actions/runs/34933608147 and https://github.com/omacom/omarchy-mac/actions/runs/34933604428 .
- **ARM Install VM is currently failing:** https://github.com/omacom/omarchy-mac/actions/runs/34933608090 . The fixture checksums and both package builds passed, then nested bubblewrap stopped with `bwrap: loopback: Failed RTM_NEWADDR: Operation not permitted` before any native scenario executed. Hosted native install/upgrade coverage is still unresolved. No host protection was relaxed and no test was silently skipped.

Earlier qualification archives built from `55542531` predate the review cleanup and are historical evidence, not publishable artifacts of the current head. No production signing secrets, host trust stores, tags or releases were changed. Publication still requires a reviewed current inventory, correct archive reuse and actual fresh/existing Mac qualification. Clients must receive the RC4 trust bootstrap before signed RC5 or edge conversion.

### CODE REVIEW

File: migrations/1789390468.sh
Comment (wesleygrimes) [inline migrations/1789390468.sh:1]: Is this migration still needed alongside `1789407944.sh`? Both populate `omarchy-mac` and verify the same primary, but their gates disagree: this one accepts any installed `omarchy-mac-keyring` (so a `20260913-1` install with only the old key passes the gate, populates, and then fails the fingerprint grep with no explanatory message), while `944` requires `>=20260914-2` and gives a clear error. `468` was never on `quattro`, so no client has completed it. Fresh clients currently run three populates (`115`, `468`, `944`). Suggest dropping this file, or documenting why both successors must coexist.

File: test/shell.d/omarchy-mac-keyring-package-install-test.sh [hunk 1]
```diff
@@ -0,0 +1,108 @@
...
+set -euo pipefail
+source "$(dirname -- "${BASH_SOURCE[0]}")/base-test.sh"
+
+# Explicit opt-in: this builds a package and initializes disposable GPG trust.
+if [[ ${OMARCHY_RUN_NATIVE_KEYRING_TEST:-0} != 1 ]]; then
+  printf 'ok - native keyring package install # SKIP set OMARCHY_RUN_NATIVE_KEYRING_TEST=1\n'
+  exit 0
+fi
+for tool in bwrap makepkg pacman gpg findmnt; do require_command "$tool"; done
+scratch_parent=${TMPDIR:-/home/naeem/.cache/codex/tmp}
```
Comment (wesleygrimes) [inline test/shell.d/omarchy-mac-keyring-package-install-test.sh]: This fallback is a developer-specific home path. Use a neutral default (e.g. `${XDG_CACHE_HOME:-$HOME/.cache}/omarchy-test` or just require `TMPDIR` to be set and fail otherwise).

File: test/shell.d/omarchy-mac-keyring-package-install-test.sh [hunk 2]
```diff
@@ -0,0 +1,108 @@
+#!/bin/bash
+set -euo pipefail
+source "$(dirname -- "${BASH_SOURCE[0]}")/base-test.sh"
+
+# Explicit opt-in: this builds a package and initializes disposable GPG trust.
+if [[ ${OMARCHY_RUN_NATIVE_KEYRING_TEST:-0} != 1 ]]; then
```
Comment (wesleygrimes) [inline test/shell.d/omarchy-mac-keyring-package-install-test.sh:6]: This is the only test that exercises the real `20260914-1` → `20260914-2` upgrade path (post_upgrade restoring the new key, old and unrelated trust preserved), and it is skipped by default and depends on a commit that only exists locally. Since this PR *is* the trust transition, either wire it into the hosted ARM job or add the upgrade scenario explicitly to the canary checklist before RC4 publication so it is not silently untested.

File: docs/arm-package-sources.md [hunk 3]
```diff
@@ -12,7 +12,7 @@ Use `omarchy update` for system upgrades. A bare `pacman -Syu` does not update t
...
+Omarchy Mac's new package-signing primary is `FBD6874D423C418DDB6D143EECE19CDDE306DBD2`, with signing subkey `D791ED0C72439D9F8757421258043B2770A25762`. Only the new fork primary ships in `omarchy-mac-keyring` version `20260914-2`; upstream keyrings remain installed. This local RC4 bootstrap candidate delivers and populates the keyring without changing the existing fork signature policy. New bootstrap configurations use the explicitly disclosed `Optional TrustAll` policy; stricter existing policies are preserved. After actual client delivery and upgrade validation, a separate signed RC5 may activate `PackageRequired DatabaseRequired TrustedOnly`. This candidate contains no strict-signing activation migration, and preparing it does not publish a release.
```
Comment (wesleygrimes) [inline docs/arm-package-sources.md:15]: Two things worth one sentence each here:

1. Why the primary changed from `F3C5AE3F…` to `FBD6874D…` (housekeeping before first publication, not a compromise).
2. That `omarchy-mac-revoked` is empty and `pacman-key --populate` does not remove keys, so machines that already installed the earlier keyring keep *both* primaries trusted, while fresh installs trust only the new one. The native test asserts exactly this, but the doc currently reads as if only the new key exists anywhere.

File: test/shell.d/omarchy-mac-keyring-successor-test.sh [hunk 4]
```diff
@@ -0,0 +1,60 @@
...
+  esac
+}
+omarchy-notification-dismiss() { :; }
+export -f pacman sudo omarchy-notification-dismiss
+
+for scenario in current newer missing older populate fingerprint; do
+  markers="$test_tmp/$scenario"
+  mkdir "$markers"
+  touch "$markers/1789316115.sh" "$markers/1789317000.sh" "$markers/1789390468.sh"
+  export TEST_VERSION=20260914-2 TEST_POPULATE_FAILURE=0 TEST_MISSING_KEY=0
```
Comment (wesleygrimes) [inline test/shell.d/omarchy-mac-keyring-successor-test.sh:38]: The `1789317000.sh` marker here models machines that already ran #436's strict-signing migration. Those machines now have `PackageRequired DatabaseRequired TrustedOnly` on `omarchy-aarch64`, and this PR deliberately preserves stricter policies, so they cannot install the unsigned RC4 packages at all. Worth calling out in the rollout notes: any dev/canary host that ran the stacked rc5 branch needs a policy reset (or reinstall) before it can act as an RC4 canary.

**Review by wesleygrimes — COMMENTED** [review]
Direction and sequencing look right: no template policy change, no strict migration, version stays rc4, and the build contract now enforces the versioned keyring dependency to match omarchy-mac/omarchy-pkgs-aarch64#28. Verified the committed keyring binaries: base `quattro` holds only `F3C5AE3F…`/`6C2597C6…`, this branch holds only `FBD6874D…`/`D791ED0C…`.

One cross-PR note: #436 still pins the old primary and subkey in `1789317000.sh`, the docs, and its test fixture, so it needs a rebase onto this before it can merge.

Inline comments cover the remaining items: the apparently redundant `1789390468.sh`, a hardcoded developer path in the native test, the upgrade test being skipped in hosted CI, a doc note on why the key changed and that upgraded hosts keep both keys, and the strict-policy canary edge case.

**Review by scottjones — COMMENTED** [review]
Reviewed against `quattro`, ran the touched suites locally, and read Wesley's review.

CI is green (7/7). All six touched/added shell suites pass here — the one `package-build-contract-test.sh` failure I hit was a missing sibling `omarchy-pkgs` checkout, environmental, and it passes once linked.

Direction and sequencing look right to me too. One improvement worth naming explicitly: in `omarchy_arm_channel_trust_fork` the fingerprint verification now runs unconditionally, where the old code only verified when the key had just been added. A keyring that already held a *wrong* key under that name would previously have been lsigned without a check. Good change.

I agree with all five of Wesley's items. Two I've sharpened in-thread:

- The skipped native upgrade test is worse than "only exists locally" — the pinned baseline commit is not in the repo at all (GitHub returns 422 for it), so the opt-in flag cannot work for anyone but the author.
- Dropping `1789390468.sh` also removes a live bug: it uses `return` at script top level, but migrations are executed rather than sourced.

Four further items inline. None of them block on their own, but the first is a rollout-ordering hazard I'd want settled before publication.

Cross-PR: confirmed #436 still pins `F3C5AE3F…` in `1789317000.sh`, so Wesley's rebase note holds.

File: migrations/1789316115.sh [hunk 5]
```diff
@@ -1,6 +1,6 @@
 echo "Trust packages signed by Omarchy Mac"
 
-readonly omarchy_mac_signing_key='F3C5AE3FCFFC738C301E30A8F0C548C0D27279F7'
+readonly omarchy_mac_signing_key='FBD6874D423C418DDB6D143EECE19CDDE306DBD2'
```
Comment (scottjones) [inline migrations/1789316115.sh:3]: This is an already-shipped migration edited in place, and its failure mode is silent — which combines badly with the publication ordering described in the PR body.

Clients that already ran `115` have a marker and are unaffected; they pick up the new key through `944`. But a client that has *not* yet run it now executes the new version: `omarchy-pkg-add omarchy-mac-keyring` with no version constraint, then a bare `grep -qF` for the new fingerprint. If the repo is still serving `20260913-1` / `20260914-1` at that moment — and the PR body says publication follows this merge — the populate succeeds with the old key, the grep fails with **no message**, `set -e` aborts `omarchy-migrate`, and every later migration never runs. That includes `944`, the one migration with a clear explanatory error.

So the failure surfaces as `omarchy update` dying with no output, and the migration that would have explained it is the one that got skipped.

Wesley flagged this silent-grep shape on `468`. It matters more here because `115` runs first and gates the rest of the queue. Suggest either giving `115` the same version gate and message as `944`, or pinning the install: `omarchy-pkg-add 'omarchy-mac-keyring>=20260914-2'`.

File: install/helpers/arm-channel.sh [hunk 6]
```diff
@@ -213,6 +213,22 @@ omarchy_arm_channel_key_fingerprints() {
 
 # Preflight has no installed-package/config/keyring side effects. The caller
 # retains this directory until applying or abandoning the captured transaction.
+# Trust the pinned new primary only inside the private transaction keyring.
```
Comment (scottjones) [inline install/helpers/arm-channel.sh:214]: The new function landed between the existing comment and the function that comment describes, so

```
# Preflight has no installed-package/config/keyring side effects. The caller
# retains this directory until applying or abandoning the captured transaction.
```

now reads as documentation for `omarchy_arm_channel_trust_fork` rather than for `omarchy_arm_channel_prepare`. Moving `trust_fork` above that comment block restores it.

File: install.sh [hunk 7]
```diff
@@ -11,7 +11,7 @@ set -euo pipefail
 readonly checkout="$(cd -- "$(dirname -- "${BASH_SOURCE[0]}")" && pwd)"
 readonly package_output="$checkout/build-output"
 readonly asahi_alarm_key="12CE6799A94A3F1B5DDFFE88F576553597FB8FEB"
-readonly omarchy_mac_key="F3C5AE3FCFFC738C301E30A8F0C548C0D27279F7"
+readonly omarchy_mac_key="FBD6874D423C418DDB6D143EECE19CDDE306DBD2"
```
Comment (scottjones) [inline install.sh:14]: After this PR the primary fingerprint is hardcoded in six places: here, `active_key` in `arm-channel.sh`, all three migrations, and `default/pacman/keyrings/omarchy-mac-trusted` — plus a literal in each of the new test files.

`install.sh` already sources `arm-channel.sh`, so at minimum those two could share one constant rather than declaring the same 40 hex characters twice. This PR is itself the evidence that rotations happen; right now performing one means a six-file grep with no single source of truth to check against.

File: build-packages.sh [hunk 8]
```diff
@@ -122,14 +122,53 @@ ensure_snapper_dependency() {
 }
 
 ensure_omarchy_mac_keyring_dependency() {
-  local pkgbuild="$1"
+  local pkgbuild="$1" minimum="20260914-2"
```
Comment (scottjones) [inline build-packages.sh:125]: This went from four lines of `sed` to ~45 lines of array and regex machinery. It is correct as far as I can tell — I confirmed the pattern does not false-positive on a neighbouring name like `'omarchy-mac-keyring-extra'` (the optional group requires an operator, so the closing quote never matches), and the new contract tests genuinely cover shared lines, trailing comments and stronger existing bounds.

Two things anyway:

1. The generality is speculative. Multiple dependencies on one line with a trailing comment is a shape the actual omarchy PKGBUILD does not have, and the tests construct it themselves to exercise the new code. A narrower change that just bumps the pin would be a lot less surface to reason about later.
2. `scan`, `token`, `suffix` are thin against the repo's full-English-name guidance — `remaining_line`, `dependency_token`, `version_constraint` would read the way the rest of this file does.

File: migrations/1789390468.sh
Comment (scottjones) [reply migrations/1789390468.sh:1]: Agreed, and dropping it also removes a live bug rather than just redundancy.

`468` uses `return 1` at script top level, but `bin/omarchy-migrate` *executes* migrations rather than sourcing them (`bash -euo pipefail "$file"`, line 94). So the failure path you describe actually looks like this:

```
$ omarchy-pkg-missing() { return 0; }; export -f omarchy-pkg-missing
$ bash -euo pipefail migrations/1789390468.sh
Populate Omarchy Mac signing trust with the new key
Install the reviewed Omarchy Mac keyring transition before activating the replacement signer.
migrations/1789390468.sh: line 7: return: can only `return' from a function or sourced script
exit=2
```

`944` correctly uses `exit 1`, so this is `468`-only.

Worth noting the suites do reach this path and pass over it: the `package` case in `omarchy-mac-keyring-transition-test.sh` sets `TEST_PACKAGE_PRESENT=0` and hits exactly this branch, but only asserts a non-zero exit and an unset marker, so the bash error on stderr goes unobserved. If `468` stays for some reason, that assertion should also pin the stderr text.

File: test/shell.d/omarchy-mac-keyring-package-install-test.sh (hunk 2 above)
Comment (scottjones) [reply test/shell.d/omarchy-mac-keyring-package-install-test.sh:6]: Seconding this, and it is a harder blocker than "exists only locally" — the pinned baseline is not in this repository at all:

```
$ git cat-file -t 61e7f31a3b47608325998cad5a1e8912f559e7cb
fatal: git cat-file: could not get object info
$ gh api repos/omacom/omarchy-mac/commits/61e7f31a3b47608325998cad5a1e8912f559e7cb
gh: No commit found for SHA: 61e7f31a3b47608325998cad5a1e8912f559e7cb (HTTP 422)
```

`git ls-remote` finds it on no ref either. So setting `OMARCHY_RUN_NATIVE_KEYRING_TEST=1` cannot work for any reviewer or CI runner — the `git show "$baseline:..."` calls fail before `makepkg` is ever reached. The only coverage of the actual `20260914-1` → `-2` transition, which is the substance of this PR, is unrunnable by anyone but the author's machine, and will stay that way even if the job is wired up.

Wiring it into the hosted ARM job therefore is not sufficient on its own. The prior `PKGBUILD`, `.install` and the three keyring files need to be committed as a fixture under `test/` (or the baseline re-pointed at a commit that is actually pushed) before the opt-in path means anything. The canary-checklist route you suggest would work as an interim, but I would not want it to be the permanent answer for the one transition this release is about.

File: migrations/1789390468.sh
Comment (malik-na) [reply migrations/1789390468.sh:1]: Addressed in 07753c84583aee5f112adc3d74bacdb300a7be0d. Removed 1789390468.sh, including its invalid top-level return. The retained 1789407944.sh successor still covers users who completed earlier migrations and requires keyring >=20260914-2. Updated the boundary and successor tests: only 115 and 944 run when pending, existing markers remain skipped, and missing/stale packages or trust failures leave the queue pending. Failure assertions now require useful diagnostics and reject the top-level return error. The focused tests pass.

File: test/shell.d/omarchy-mac-keyring-package-install-test.sh (hunk 1 above)
Comment (malik-na) [reply test/shell.d/omarchy-mac-keyring-package-install-test.sh]: Addressed in 07753c84583aee5f112adc3d74bacdb300a7be0d. Removed the developer-specific home path. Scratch now uses OMARCHY_TEST_TMPDIR, then TMPDIR, then /var/tmp, and still rejects tmpfs/ramfs or an unidentifiable backing filesystem before creating test files. The hosted guest explicitly supplies /var/tmp.

File: test/shell.d/omarchy-mac-keyring-package-install-test.sh (hunk 2 above)
Comment (malik-na) [reply test/shell.d/omarchy-mac-keyring-package-install-test.sh:6]: The reproducibility and CI wiring changes are in 07753c84583aee5f112adc3d74bacdb300a7be0d, but hosted execution is not yet passing.

Committed the previous public PKGBUILD, install script and three keyring files under test/fixtures/omarchy-mac-keyring-20260914-1/, with checksums and provenance. The test no longer calls git show or needs the unpublished commit. All three native scenarios passed locally from a source-only directory with no .git. The existing hosted ARM guest now explicitly enables the test, uses disk-backed scratch and runs it without networking.

The new hosted run verified all five fixture checksums and built both package versions, then stopped before the native scenarios with `bwrap: loopback: Failed RTM_NEWADDR: Operation not permitted`. This is an unresolved nested-containment setup failure; it is not a hosted package-hook pass. Run: https://github.com/omacom/omarchy-mac/actions/runs/34933608090 . Keeping this conversation open until the test actually executes successfully in CI. No host protection has been relaxed or test silently skipped.

File: docs/arm-package-sources.md (hunk 3 above)
Comment (malik-na) [reply docs/arm-package-sources.md:15]: Addressed in 07753c84583aee5f112adc3d74bacdb300a7be0d, in docs/arm-package-sources.md under “RC4 trust bootstrap qualification.” It now states that the primary replacement was pre-activation housekeeping, not a compromise response; the shipped revoked file is empty; and population preserves already installed old/unrelated trust. Fresh hosts receive only the new fork primary. Shipping only the new certificate is explicitly distinguished from removing old trust on upgraded machines.

File: test/shell.d/omarchy-mac-keyring-successor-test.sh (hunk 4 above)
Comment (malik-na) [reply test/shell.d/omarchy-mac-keyring-successor-test.sh:38]: Addressed in 07753c84583aee5f112adc3d74bacdb300a7be0d. The rollout notes explicitly exclude an already-strict development host from unsigned-RC4 canary qualification and direct qualification to a clean disposable baseline with the supported RC4 policy. No automatic policy reset or marker deletion is introduced. The successor fixture also now says that its obsolete markers test marker handling only, not whether an unsigned transaction can succeed under strict signing.

File: migrations/1789316115.sh (hunk 5 above)
Comment (malik-na) [reply migrations/1789316115.sh:3]: Addressed in 07753c84583aee5f112adc3d74bacdb300a7be0d using the explicit installed-version gate. 115 now requires >=20260914-2 before its explicit populate call and reports missing/outdated packages, population failure and a missing expected fingerprint. The successor remains responsible for clients whose 115 marker already exists.

Kept the helper argument as the plain package name: omarchy-pkg-add currently queries pacman -Q/-Si with that argument, so passing a dependency constraint is not a supported drop-in fix. The gate also catches the helper returning success after skipping an unavailable package. Package installation can still run the package's own scriptlets; this is a gate before the migration's explicit trust operations.

A before/after test with installed 20260913-1 showed the original exiting after two explicit trust calls without a useful explanation; the new migration exits before those calls with the required and installed versions. Tests through the actual migration runner and real package helpers cover missing/unavailable and outdated packages, successful repair, fingerprint/population failure, and both migration markers remaining pending on failure. The documentation now states the package-before-migration ordering.

One detail: the local 20260914-1 fixture contains both primaries, so that version is not an old-only certificate example. It is nevertheless below the required package revision and is explicitly rejected by the new gate.

File: install/helpers/arm-channel.sh (hunk 6 above)
Comment (malik-na) [reply install/helpers/arm-channel.sh:214]: Addressed in 07753c84583aee5f112adc3d74bacdb300a7be0d. Moved the two-line preflight comment immediately above omarchy_arm_channel_prepare. trust_fork retains its own private-keyring comment. This is a comment-only change; helper behavior is unchanged.

File: install.sh (hunk 7 above)
Comment (malik-na) [reply install.sh:14]: Deferred in this cleanup; no shared-constant refactor was made. The current installer/helper/package pins agree, and their consistency checks remain in place. A shared active pin for the installer and helper is reasonable, while historical migrations should retain their explicit expected fingerprint. The owner asked to keep this pass focused on the migration failures, reproducible upgrade coverage and rollout documentation. Leaving this suggestion open rather than presenting it as implemented.

File: build-packages.sh (hunk 8 above)
Comment (malik-na) [reply build-packages.sh:125]: Deferred in this cleanup. build-packages.sh is unchanged in 07753c84583aee5f112adc3d74bacdb300a7be0d, including the variable names. The existing behavior preserves stronger constraints and neighboring dependencies; the current contract tests remain. I agree that the parser's scope and naming deserve a separate simplification pass, but the owner asked to keep this cleanup focused on the concrete migration/test/rollout issues. Neither parser simplification nor the naming suggestion is claimed as addressed; leaving this conversation open.

File: test/shell.d/omarchy-mac-keyring-package-install-test.sh (hunk 2 above)
Comment (malik-na) [reply test/shell.d/omarchy-mac-keyring-package-install-test.sh:6]: Final status at `1b117466`: the previous public `20260914-1` recipe, install script, keyring files and checksums are committed under `test/fixtures/`, so the real upgrade test is reproducible without local git history. The direct native test passed in the Omarchy Mac Test Arch ARM VM for all three scenarios: uninitialized trust, fresh install preserving unrelated trust, and `20260914-1` → `20260914-2` restoring the new key while preserving old and unrelated trust. We also corrected the coverage claim: nested `bwrap` inside the hosted Ubuntu/nspawn job failed before package-hook assertions, so `1b117466` removes that unreliable wiring and documents the direct Arch ARM test plus the explicit RC4 canary checklist instead. Exact-head CI is green, including the Install VM job for its actual package-source/install/idempotency scope.

**Comment by malik-na** [comment]
Review follow-up at `1b117466`:

Resolved as implemented and verified: redundant/broken `1789390468` removal; neutral disk-backed scratch selection; committed reproducible `20260914-1` fixture; key-rotation/trust-preservation documentation; strict-policy canary exclusion; the `1789316115` version/error gate; and the misplaced helper comment.

The native keyring test passed directly in the Arch ARM UTM VM for uninitialized, fresh-install, and real `20260914-1` → `20260914-2` upgrade scenarios. The Ubuntu/nspawn workflow no longer claims nested native-keyring coverage; the RC4 qualification checklist remains explicit. All exact-head CI checks are green.

Left open intentionally as non-blocking follow-ups: consolidating the active fingerprint constant and simplifying/renaming the dependency parser.

**Comment by malik-na** [comment]
## Required merge and rollout order

The PR dependency order is:

1. `omarchy-pkgs-aarch64#28` — merged.
2. This RC4 bootstrap PR (`omacom/omarchy-mac#444`).
3. The strict-signing RC5 PR (`omacom/omarchy-mac#436`).

Publication gates are part of that order; merging the source PRs alone is not sufficient:

1. Publish `omarchy-mac-keyring 20260914-2` and the RC4 packages built from the reviewed `#444` head.
2. Confirm the repository exposes the expected versions and that `omarchy` requires `omarchy-mac-keyring>=20260914-2`.
3. Complete the clean-install and upgrade checks in the Arch ARM UTM VM.
4. Merge `#444`, then verify a real client update completes the bootstrap migrations. Keep issue `#446` open until this deployed-path check succeeds.
5. Rebase/update `#436` on the merged `#444`, including replacing any stale fingerprint assumptions.
6. Build, sign, publish, and canary-test the RC5 package set and repository database.
7. Merge `#436` only after a bootstrapped client successfully consumes those signed artifacts under the intended strict policy.

In short: **#28 → publish and verify RC4 bootstrap artifacts → #444 → verify deployed bootstrap → prepare and canary signed RC5 → #436**.

Do not merge `#444` while its required keyring package is unavailable, and do not merge `#436` before the `#444` bootstrap has been deployed and verified.

File: test/fixtures/omarchy-mac-keyring-20260914-1/omarchy-mac-trusted [hunk 9]
```diff
@@ -0,0 +1,2 @@
+F3C5AE3FCFFC738C301E30A8F0C548C0D27279F7:4:
+FBD6874D423C418DDB6D143EECE19CDDE306DBD2:4:
```
Comment (scottjones) [inline test/fixtures/omarchy-mac-keyring-20260914-1/omarchy-mac-trusted:2]: Small follow-up, not a blocker, and it does not affect the migration coverage — which is solid.

This fixture is `20260914-1` with both primaries, but that version was never on `quattro`. What shipped, and what installed clients actually have, is `20260913-1`, whose `omarchy-mac-trusted` lists only the old primary:

```
$ git show origin/quattro:build-inputs/omarchy-mac-keyring/PKGBUILD | grep -E '^pkgver|^pkgrel'
pkgver=20260913
pkgrel=1
$ git show origin/quattro:default/pacman/keyrings/omarchy-mac-trusted
F3C5AE3FCFFC738C301E30A8F0C548C0D27279F7:4:
```

So the native test exercises *both-keys → new-key-only*, while real hosts will do *old-key-only → new-key-only*.

The practical effect is narrow but it is the reason I am mentioning it: because the fixture already trusts the new primary, line 84 has to `pacman-key --delete "$key"` to manufacture something for `post_upgrade` to restore. That is a simulation of the upgrade condition rather than the condition itself — on a real host the new primary was simply never present, and "establish a key that was never there" is a slightly different path through the scriptlet than "restore a key that was just removed."

You already noted the fixture is not an old-only certificate example, but tied it to the migration version gate; this is about the package-hook path the native test covers, which is a different mechanism.

Cheap fix if you think it is worth it: commit the five `20260913-1` sources from `quattro` as a second fixture directory the same way, and add a third upgrade scenario. That would make the one test of the actual shipping upgrade path model the actual shipping upgrade path, with no manual key deletion. Keeping the existing `20260914-1` scenario alongside it still has value for the both-keys intermediate state.

Entirely your call on whether this lands before RC4 or after — the canary checklist would also catch it on real hardware.

**Review by scottjones — COMMENTED** [review]
Re-reviewed at `1b117466`. Verified the fixes rather than taking them on trust:

- `115`'s new gate behaves as described — against an installed `20260913-1` it now exits `1` with `Omarchy Mac keyring 20260914-2 or newer is required (installed: 20260913-1); …` and makes no `pacman-key` call, where the previous version aborted the whole queue silently. The nine-scenario queue test covers `old-only` and `install-old`, which is exactly the publication-ordering hazard.
- `1789390468.sh` is gone, and both suites now assert against its top-level `return` error rather than just a non-zero exit.
- Fixture checksums verify clean; the test runs with no `.git` present.
- The preflight comment is back above `omarchy_arm_channel_prepare`.

All eight suites this PR touches pass here, and CI is green on the exact head. For the two failures in my full-suite run (`install-vm-source-isolation`, `plymouth-set`): they reproduce identically on unmodified `quattro` under the same environment, so they are my local setup, not this branch. The five test files this PR adds all pass.

I also want to note the handling of the hosted-CI item: removing the nspawn wiring and adding assertions that stop that job from re-claiming the coverage is a better outcome than leaving it in place and calling it covered. Same for marking the `install.sh` constant and the `build-packages.sh` parser as deferred rather than folding them in silently — no objection to either deferral, neither was a correctness issue.

One small follow-up inline, not a blocker.

File: test/fixtures/omarchy-mac-keyring-20260914-1/omarchy-mac-trusted (hunk 9 above)
Comment (malik-na) [reply test/fixtures/omarchy-mac-keyring-20260914-1/omarchy-mac-trusted:2]: Addressed in `a61ee873`. Added a byte-for-byte, checksummed `20260913-1` fixture from `origin/quattro@04862a02` and changed the native test to exercise the actually shipped old-primary-only package. The upgrade scenario now starts with independently reinitialized disposable trust, proves the new primary was never present, installs `20260913-1`, and performs the real pacman upgrade to `20260914-2`; the targeted `pacman-key --delete` simulation is gone.

QA transferred the exact working-tree files into the Arch ARM UTM VM with matching archive hashes. The native test exited 0: uninitialized and fresh-install scenarios still pass, and the shipped `20260913-1` → `20260914-2` path adds the new primary while preserving the old and unrelated trusted keys. Fixture files also compare byte-for-byte with `quattro`, their manifest verifies, syntax/diff checks pass, and independent review found no blocking issue. Production signing behavior is unchanged.

Comment (scottjones) [reply test/fixtures/omarchy-mac-keyring-20260914-1/omarchy-mac-trusted:2]: Verified at `a61ee873`:

- All five files under `test/fixtures/omarchy-mac-keyring-20260913-1/` are byte-identical to `origin/quattro@04862a02c`, and `SHA256SUMS` checks clean.
- Ran the native test here with `OMARCHY_RUN_NATIVE_KEYRING_TEST=1` on disk-backed scratch: all three scenarios pass, including the shipped `20260913-1` → `20260914-2` upgrade with the manual `pacman-key --delete` gone.
- CI is green on the exact head.

That closes this thread from my side. One small leftover: `test/fixtures/omarchy-mac-keyring-20260914-1/` is still in the tree but nothing references it now (`git grep omarchy-mac-keyring-20260914-1` is empty). Replacing rather than adding alongside is fine, but the unused directory should either be removed or get a scenario that uses it, so the next person does not have to work out which fixture is live. Not a blocker.

**Review by wesleygrimes — APPROVED** [review]
Approved at `a61ee873`. The RC4 bootstrap logic and migration sequencing look sound, and exact-head CI is green. No source blocker from my review.

Approval assumes the documented rollout order is followed: publish and verify `omarchy-mac-keyring >= 20260914-2` and the matching RC4 artifacts first, complete fresh-install and real-client upgrade validation, then merge this PR. After deployment, #436 must be rebased onto this head before strict signing is enabled.
